### ISSUE TITLE
Algorithm 1 should return both the HCGM with Q-variables and the collapsed model

### ISSUE THREAD

**Issue by djinnome** [y0-causal-inference/y0#239]
Because Algorithm 2 needs to traverse a graph containing both Q variables and subunit-level variables, the HCGM should be returned as part of Algorithm 2.  

One way to accomplish this is to split algorithm 1 into two parts:  `promote_subunit_variables`, which adds Q-variables, and `remove_subunit_variables` which removes subunit variables.  Importantly, the output of `promote_subunit_variables` should be the input to `remove_subunit_variables`

**Comment by djinnome** [comment]
This should work for figure 2, but we should test that our intuition works for more sophisticated examples

**Comment by djinnome** [comment]
Furthermore, Algorithm 2 must take as input the causal query, composed of the intervention and the outcome (and potentially any conditioned variables) so that we know which Q variables must be augmented.

Now some variables that are not in the query need to be augmented because they are needed to generate the Q-variables in the query

**Comment by adamrupe** [comment]
I've implemented something similar to your suggestion for Algorithm 1. 
I've made a `convert_to_HCGM` function that is the equivalent of your suggested `promote_subunit_variables`. I then updated `collapse_HCM` to first call `convert_to_HCGM` and then remove the subunit variables from the generated HCGM. This code follows the paper's pseudo code much better. There is also a bool flag in the new `collapse_HCM` with the option to also return the intermediate HCGM.
